---
title: "Sequence Data Processing"
author: "Melanie Smith"
date: "7 December 2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = TRUE,
    message = FALSE,
    warning = FALSE,
    cache = FALSE,
    fig.align = "center"
)
```

This RScript is the second in the collaboration with Anna and Burnd.  
Here I will inspect the sequencing data derived from the STOP and LMH (Anna's Iron Infusion Project) placentas.  


```{r include=FALSE}

###################################################
#                 Load Libraries                  #
###################################################

# Data manipulation
library(dplyr)
library(magrittr)
library(reshape2)
library(readr)
library(tibble)
library(readxl)

# Visualisation
library(ggplot2)
library(ggrepel)
library(RColorBrewer)
library(knitr)
library(kableExtra)
library(plotly)


###################################################
#              Set ggplot theme                   #
###################################################

theme_set(theme_bw() +
            theme(text = element_text(size=18)) +
  theme(axis.text.x = element_text(colour = "black",
                                   face = "bold"),
        axis.text.y = element_text(colour = "black",
                                   face = "bold")) +
  theme(axis.title.x = element_text(colour = "black",
                                   face = "bold"),
        axis.title.y = element_text(colour = "black",
                                   face = "bold")))

###################################################
#                   Directories                   #
###################################################

projectDir <- "/home/smit1924/ironReceptor_placenta"

input_raw_counts <- file.path(projectDir, "rawData/SAGCQA0492_readCounts.txt")
input_filename_convert <- file.path(projectDir, "rawData/20230405_jointMetaData.csv")
input_filename_flinders_placentas_bloods <- file.path(projectDir, "rawData/flinders_placentas_and_bloods_17_11_2022_aw.xlsx")
input_filename_lmh_placentas_bloods <- file.path(projectDir, "rawData/20230405_LMH_metaData.csv")
input_filename_fourthSubsetSTOP <- file.path(projectDir, "cleanData/fourthSubsetSTOP_UR.xlsx")

outputDir <- file.path(projectDir, "cleanData")
dataFolder <- file.path(projectDir, "dataFolder")


###################################################
#             User Defined Functions              #
###################################################

# Negate the magrittr %in% function
`%notin%` <- Negate(`%in%`)
# REDCap import
source(file.path(projectDir, "RScripts", "redcap_api_import.R"))
```

# Import sample metadata

```{r}
# import combined file name to study ID conversion table
filename_convert <- read_csv(input_filename_convert, col_names = TRUE)
# import STOP metadata
flinders_placentas_and_bloods_20221117 <- read_excel(input_filename_flinders_placentas_bloods)
# import LMH metadata
LMH_metaData_20230405 <- read_csv(input_filename_lmh_placentas_bloods) %>%
  # keep only columns that match the flinders metadata
  dplyr::select(., "ULN", "studyno", "f1_age", "f9_hgt", "f9_wgt", "f9_bmi", "f4_ethnicity", "f22_birthwgt", "f21_est_blood_loss", "f22_sex", "f22_gestw_del", "f21_del_mode")

# import some extra STOP metadata
fourthSubsetSTOP <- read_excel(input_filename_fourthSubsetSTOP)

```

## combine metadata
```{r}

temp1 <- fourthSubsetSTOP %>%
  dplyr::left_join(., filename_convert, by =  "samplename") %>%
  dplyr::select(., samplename, f22_sex, ULN)  %>%
  dplyr::filter(!is.na(ULN))

female <- dplyr::filter(temp1, f22_sex == "Female")
male <- dplyr::filter(temp1, f22_sex == "Male")

```

# Import raw count data

```{r}
# import counts
rawCounts <- read.delim(file = input_raw_counts) %>%
  as.data.frame() %>%
  # remove the 'undetermined' column
  dplyr::select(., -X.scratch.user.smit1924.annaWaverIronProject.aligned_data.Undetermined_GRCh38_Aligned.sortedByCoord.out.bam) %>%
  # remove the sequencing blank `SAGCFN_22_01736`
  dplyr::select(., -X.scratch.user.smit1924.annaWaverIronProject.aligned_data.SAGCFN_22_01736_S71_GRCh38_Aligned.sortedByCoord.out.bam) %>%
  # move the gene IDs to the rownames
  tibble::column_to_rownames("Geneid")

# tidy up the column names
colnames(rawCounts) <- gsub("X.scratch.user.smit1924.annaWaverIronProject.aligned_data.|_S.*|
                                 _GRCh38_Aligned.sortedByCoord.out.bam",
                                 "",
                                 colnames(rawCounts))
```