---
title: "tpm_calc_combinedSex"
author: "Melanie Smith"
date: "22 March 2024"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = TRUE,
    message = FALSE,
    warning = FALSE,
    cache = FALSE,
    fig.align = "center",
    results = "hide"
)

```

This script converts the counts generated by `featureCounts` into transcripts per million (TPMs).  
Conversion to TPMs is required before we can compare the bulk counts with scRNAseq counts.  
In this case, the comparison will be used for cell type deconvolution.  

```{r}
library(IRanges)
library(GenomicFeatures)
library(edgeR)
library(dplyr)
library(stringr)
# library(tidyverse)
library(readxl)
library(readr)
library(magrittr)

# set project directory
projectDir <- "/home/smit1924/ironReceptor_placenta"

genome_gtf_infile <- file.path(projectDir, "rawData/gencode.v29.annotation.gtf")
gencodev29_gff_inFile <- file.path(projectDir, "rawData/gencode.v29.chr_patch_hapl_scaff.annotation.gff3")

input_file_counts_female <- file.path(projectDir, "rawData/20240315_ironProject_female_readCounts.txt")
input_file_counts_male <- file.path(projectDir, "rawData/20240315_ironProject_male_readCounts.txt")
input_file_metadata <- file.path(projectDir, "cleanData/ironReceptor_metadata.csv")

outdir <- file.path(projectDir, "cleanData")
# output_female_cibersort_file <- file.path(outdir, paste0(expt_name, "_bulkCountMatrix_female.txt"))
# output_male_cibersort_file <- file.path(outdir, paste0(expt_name, "_bulkCountMatrix_male.txt"))
output_all_cibersort_file <- file.path(outdir, "tpm_bulkCountMatrix_all.txt")
output_all_cibersort_file_gene_symbol <- file.path(outdir, "tpm_bulkCountMatrix_all_gene_symbol.txt")

# dir.create(outdir)

# set max digits
options(digits=3)

# set filtering criteria
filterCPM <- 2
numSamples <- 5
```

Before I can transform my raw counts to TPMs I need to calculate the gene lengths from the original .gtf file.   

```{r}

# Try it
# Specifics of calculating rpkm: https://support.bioconductor.org/p/64585/

# 1. Get gene lengths
# https://www.biostars.org/p/83901/
# Or using GTFtools, but keep chromosome limitations in mind:
# http://genomespot.blogspot.com/2019/01/using-gtf-tools-to-get-gene-lengths.html
# GTFtools offers flexibility about how to calculate the gene length (median, mean, etc.)

# Note, this is the gtf used in the RNA-Seq pipeline to annotatate hg38
txdb <- GenomicFeatures::makeTxDbFromGFF(file = file.path(genome_gtf_infile), format = "gtf")
# Collect exons per gene id
exon_list_per_gene <- GenomicFeatures::exonsBy(txdb, by = "gene")
exon_list_per_gene


```

```{r}

# For each gene, reduce all exons to a set of non-overlapping exons, calculate their length (widths) and sum them
exonic_gene_sizes <- 
  sum(IRanges::width(IRanges::reduce(exon_list_per_gene))) %>% 
  tibble::as_tibble(rownames = NA) %>% 
  tibble::rownames_to_column(var = "gene") %>%
  dplyr::select(., gene, length = value)
exonic_gene_sizes$gene %>% length()
table(is.na(exonic_gene_sizes$length))

```

## Import metadata

```{r}
## Metadata
# import metadata
metadata <- read_csv(file = input_file_metadata, col_names = TRUE)
male <- dplyr::filter(metadata, f22_sex == "Male") %>%
  dplyr::select(., samplename)
female <- dplyr::filter(metadata, f22_sex == "Female") %>%
  dplyr::select(., samplename)
```

## Import count files
### Import counts (female)

```{r import gene counts female}

# import the counts table
# counts are annotated to GRCh38
rawCounts_female <- read.delim(file = file.path(input_file_counts_female)) %>%
  as.data.frame()
# tidy up the column names
colnames(rawCounts_female) <- gsub("X.scratch.user.smit1924.ironReceptor_placenta.fastqProcessing.femaleOutput.aligned_data.|
                                   _GRCh38_Aligned.sortedByCoord.out.bam|_S.*",
                                 "",
                                 colnames(rawCounts_female))
# remove the genes from the PAR_Y (all have zero counts)
rawCounts_female %<>% dplyr::filter(., !grepl("PAR_Y",Geneid))
# remove the number after the period
#rawCounts_female %<>% tidyr::separate(., col = Geneid, into = c("ensembl", "right")) #%>%
 # dplyr::select(., -right) %>%
 # tibble::column_to_rownames("ensembl")

dim(rawCounts_female)

```

### Import counts (male)

```{r import gene counts male}

# import the counts table
# counts are annotated to GRCh38
rawCounts_male <- read.delim(file = file.path(input_file_counts_male)) %>%
  as.data.frame()
# tidy up the column names
colnames(rawCounts_male) <- gsub("X.scratch.user.smit1924.ironReceptor_placenta.fastqProcessing.maleOutput.aligned_data.|
                                   _GRCh38_Aligned.sortedByCoord.out.bam|_S.*",
                                 "",
                                 colnames(rawCounts_male))
# remove the genes from the PAR_Y (all have zero counts)
rawCounts_male %<>% dplyr::filter(., !grepl("PAR_Y",Geneid))
# remove the number after the period
# rawCounts_male %<>% tidyr::separate(., col = Geneid, into = c("ensembl", "right")) %>%
#   dplyr::select(., -right) %>%
#   tibble::column_to_rownames("ensembl")

dim(rawCounts_male)

```

## Combine counts data into a single df

```{r combine counts, echo=FALSE}

# test to make sure the rownames are the same
identical(rownames(rawCounts_male), rownames(rawCounts_female))

# combine the counts tables
rawCounts_combinedSex <- dplyr::left_join(rawCounts_male,
                                             rawCounts_female,
                                             by = "Geneid")
dim(rawCounts_combinedSex)

# replace the file names with the samplename before moving to the DGEList steps
rawCounts_combinedSex %<>% tibble::column_to_rownames("Geneid") %>%
  dplyr::select(., metadata$ULN) %>% # mske sure the columns are in the same order as the metadata
  setNames(metadata$samplename) %>% # replace the column names
  tibble::rownames_to_column("gene_id") # rename the gene column to match Nick's code
```

```{r}

# 2. Process raw count data in edgeR again, as per melanieRawCountsProcessing_v2.0.3.rmd
# Add gene length to expression data
# at this point ensembl IDs have the number after the period
raw_data <- rawCounts_combinedSex %>%
  dplyr::mutate(Length = exonic_gene_sizes$length[match(gene_id, exonic_gene_sizes$gene)])

raw_data <- raw_data %>% 
  tibble::column_to_rownames("gene_id")
head(raw_data)
dim(raw_data) # 58676, 71

y <- edgeR::DGEList(counts = raw_data %>% dplyr::select(-Length),
             genes = raw_data %>% dplyr::select(Length))

y
y$genes # A Length column is present 
dim(y) # 58676 70

head(y$counts) # Gene names are set as rownames
y$samples$lib.size #  13216072 24873717  5827620 28117960 16715504 15556461 16877794 14439137 16372317 15753821 18368144 21119494 13001262 17483334 22362594 15488048 33209766 21369724 22318368 18298493 21983492 12248665 18617780 17575963 23851628 13971600 17426920 18961157 16903398 14238019 17726594 20444573 12605154 16996226 17694726 13290688 14421957 17778288 11736608 16000694 16356669 15417494 18927236 12232010 18638451 13900253 15150842 12304767 14208510 13987704 15318821 11968086 20155631 14368589 13127683  9801206 21727295 13563049 16075098 15172349 10849351 14137921 10517772  9749547 14017279 10924164 15426458 18033977 17426099 15753299


# Identify genes expressed at > 1 cpm in at least one sample
keep <- rowSums(cpm(y) > filterCPM) >= numSamples
table(keep) # Keeping 14082 genes

# Apply filter
y <- y[keep, , keep.lib.sizes = FALSE]
dim(y) # 14254 70
y$samples$lib.size # 13192604 24847264  5815027 28072657 16679675 15521873 16843285 14406246 16345385 15726053 18336026 21091668 12979301 17456950 22332063 15460783 33153794 21332231 22283680 18264166 21953143 12209011 18589985 17550450 23810155 13947149 17399791 18931583 16875973 14207623 17699832 20416717 12586673 16974546 17668132 13254703 14391646 17743996 11715862 15962964 16318879 15392783 18874033 12202404 18591988 13870655 15105766 12277747 14186291 13956478 15284945 11936694 20106937 14338569 13084877  9781182 21670915 13524866 15997689 15135864 10810937 14095721 10459871  9724215 13962154 10902526 15399737 17991020 17390313 15730137

# TMM normalisation 
y$samples$norm.factors # currently all normalisation factors == 1
y <- calcNormFactors(y, method = "TMM")
# print the normalisation factors
y$samples$norm.factors # 1.070 0.831 1.083 0.944 1.112 1.121 1.058 1.130 0.943 0.935 0.985 0.913 0.937 0.879 0.932 0.929 0.964 0.990 0.901 0.942 0.808 1.380 0.915 0.898 0.978 0.981 0.910 0.902 0.908 1.076 0.910 0.910 0.915 0.778 0.863 1.184 0.892 0.946 0.975 1.048 0.910 0.884 1.197 1.074 1.095 1.046 1.149 1.075 0.963 1.172 0.979 1.017 1.114 0.908 1.142 0.920 1.041 1.074 1.385 0.868 1.226 1.186 1.130 1.096 1.549 0.979 0.886 1.022 0.868 0.874

# Now calculate RPKM: https://support.bioconductor.org/p/64585/
rpkm <- edgeR::rpkm(y)

# And TPM: https://www.biostars.org/p/388584/#394579
tpm <- t( t(rpkm) / colSums(rpkm) ) * 1e6
# These TPM values now incorporate the TMM factors

```

## lets add the gene name information
### Import gencode V32 gff3 file

```{r}

# import the gencodev32_gff_inFile file
all_gencode_v29 <- rtracklayer::import(gencodev29_gff_inFile)
# this file contains more information than we need here
# subset out only the columns we need
gene_data <- data.frame(ensembl_gene_id = all_gencode_v29@elementMetadata$gene_id,
                        hgnc_symbol = all_gencode_v29@elementMetadata$gene_name,
                        seqnames = all_gencode_v29@seqnames)
# we're left with multiple identical rows, keep one each
gencode_v29_gene_id_symbol_chr_biotype <- gene_data %>%
  dplyr::distinct(, .keep_all = TRUE)

```

### add the gene info to the tpms

```{r}
# join the counts with the gene info
tpm_gene_id_symbol_chr_biotype <- data.frame(tpm) %>%
  tibble::rownames_to_column("ensembl_gene_id") %>%
  dplyr::left_join(., gencode_v29_gene_id_symbol_chr_biotype, by = "ensembl_gene_id") %>%
  dplyr::select(colnames(gencode_v29_gene_id_symbol_chr_biotype), colnames(data.frame(tpm)))

# save a copy of the tpm counts with the gene info
tpm_gene_id_symbol_chr_biotype %>%
  # remove the number after the period
  tidyr::separate(., col = ensembl_gene_id, into = c("gene", "right")) %>% 
  dplyr::select(., -right) %>%
  # save TPM count file for CIBERSTORTx
  write.table(file = output_all_cibersort_file,
              quote = FALSE,
              sep = "\t",
              col.names = TRUE,
              row.names = FALSE)

# save a copy of the tpm counts with only the hgnc gene symbol
tpm_gene_id_symbol_chr_biotype %>%
  # remove the ensembl gene id and the chromosome name
  dplyr::select(., -ensembl_gene_id, -seqnames) %>%
  # save TPM count file for CIBERSTORTx
  write.table(file = output_all_cibersort_file_gene_symbol,
              quote = FALSE,
              sep = "\t",
              col.names = TRUE,
              row.names = FALSE)

```

## Session information

```{r session info}
sessionInfo()
```